ext/opcache/jit/zend_jit.c

```diff
@@ -2724,33 +2736,34 @@ void ZEND_FASTCALL zend_jit_hot_func(zend_execute_data *execute_data, const zend
 static int zend_jit_setup_hot_counters(zend_op_array *op_array)
 {
 	zend_op *opline = op_array->opcodes;
+	zend_op *end;
 	zend_jit_op_array_extension *jit_extension;
-	zend_cfg cfg;
-	uint32_t i;
-
-	if (zend_jit_build_cfg(op_array, &cfg) != SUCCESS) {
-		return FAILURE;
-	}
 
 	jit_extension = (zend_jit_op_array_extension*)zend_shared_alloc(sizeof(zend_jit_op_array_extension) + (op_array->last - 1) * sizeof(void*));
 	jit_extension->counter = &zend_jit_hot_counters[zend_jit_op_array_hash(op_array) & (ZEND_HOT_COUNTERS_COUNT - 1)];
-	for (i = 0; i < op_array->last; i++) {
-		jit_extension->orig_handlers[i] = op_array->opcodes[i].handler;
-	}
-	ZEND_SET_FUNC_INFO(op_array, (void*)jit_extension);
+	jit_extension->orig_enter_handler = NULL;
 
 	while (opline->opcode == ZEND_RECV || opline->opcode == ZEND_RECV_INIT) {
 		opline++;
 	}
 
-	opline->handler = (const void*)zend_jit_func_counter_handler;
+	if (opline->opcode != ZEND_LOOP) {
+		jit_extension->orig_enter_handler = opline->handler;
+	}
 
-	for (i = 0; i < cfg.blocks_count; i++) {
-		if ((cfg.blocks[i].flags & ZEND_BB_REACHABLE) &&
-		    (cfg.blocks[i].flags & ZEND_BB_LOOP_HEADER)) {
-		    op_array->opcodes[cfg.blocks[i].start].handler =
+	ZEND_SET_FUNC_INFO(op_array, (void*)jit_extension);
+
+	if (opline->opcode != ZEND_LOOP) {
+		opline->handler = (const void*)zend_jit_func_counter_handler;
+	}
+
+	end = op_array->opcodes + op_array->last;
+	while (opline != end) {
+		if (opline->opcode == ZEND_LOOP) {
```

Comment:
Won't we miss the FE_FETCH case now?

Reply:
Right. I'm going to restore FE_FETCH handling a bit later.
In general, we may simple add LOOP instruction after FE_FETCH, but it would be great to avoid this.

Reply:
Okay, in that case this change looks fine to me :)

Do you have any quick numbers on whether this affects bench.php negatively?

Reply:
I added TODO comment.

Reply:
> Okay, in that case this change looks fine to me :)
> 
> Do you have any quick numbers on whether this affects bench.php (without JIT) negatively?

The patch increases number of retired CPU instructions on bench.php (measured with callgrind). This increase caused by additional virtual instruction dispatch on each loop iteration. Actually, two more CPU instructions on each loop iteration.